tools: add the doc of handling sharding DDL lock manually (#755)
pingcap/docs: tools/handle-ddl-lock.md

---
title: Handling the Sharding DDL Lock Manually

> summary: Learn to manually unlock the sharding DDL lock.

Reviewer: Should we call it "Troubleshooting Sharding DDL Locks"?  'Handling' is a bit ambiguous: it could be handling large amounts of data etc.  This sounds like break/fix? (troubleshooting)

category: tools
---

# Handling the Sharding DDL Lock Manually

This document shows how to manually unlock the sharding DDL lock in different cases.

> Generally, the sharding DDL lock synchronization of DM can be completed automatically. But when some abnormality happens, you need to use `unlock-ddl-lock`/`break-ddl-lock` to unlock the abnormal DDL lock manually.

Reviewer: Can we move this to the top before "Handling the sharding DDL lock manually"?  I did not know what a sharding lock was, so logically an intro to that helps before manual handing.  Maybe as a small tweak:
"The Data Migration tool uses a sharding DDL lock to ensure operations are applied in the correct order.  This locking mechanism occurs automatically, but in some abnormal cases you may need to perform manual operations such as force-releasing the lock."

The possible causes of an abnormality include:

- Some DM-workers go offline

> - Some DM-worker restarts (or is unreachable temporarily)

Reviewer: "Some DM-worker" is mixing singular/plural.  If the condition is true for single, we should remove plural for readability:
- A DM-worker goes offline
- A DM-worker restarts (or is temporarily unreachable)

- DM-master restarts

> **Note:** You can use `unlock-ddl-lock`/`break-ddl-lock` only when you are definitely clear about the possible impacts brought by this command and you can accept the impacts.

## Case one: some DM-workers go offline

> Before DM-master tries to automatically unlock the sharding DDL lock, all the DM-workers need to receive the sharding DDL event. If the sharding DDL operation is already in the synchronization process, and some DM-workers have gone offline and are not to be restarted, the sharding DDL lock cannot be automatically synchronized and unlocked since not all the DM-workers can receive the DDL event.

Reviewer: Before *the* DM-master..

If you do not need to make some DM-workers offline in the synchronization process of the sharding DDL operation, a better policy is to use `stop-task` to stop the running task first, then make some DM-workers offline, and finally use `start-task` and the **new task configuration** that does not contain the already offline DM-worker to restart the task.

> If the owner has finished the DDL operation but other DM-workers have not skipped this DDL operation, the owner has gone offline. For the solution, see [Case two: some DM-worker restarts](#case-two-some-dm-worker-restarts-or-is-unreachable-temporarily).

Reviewer: This sentence does not quite make sense to me.  It would be helpful to highlight the cause/effect:
Is it:
"If the owner goes offline before the DDL operation is complete **then** the other DDL workers will block waiting..." ?
"If the owner goes offline before propagating XYZ state to other DM-workers **then** ..".?
"if the owner finishes the DDL operation but other DM-workers skip the DDL operation, then the owner will be forced into offline state" ?

> If the owner has finished the DDL operation but other DM-workers have not skipped this DDL operation, the owner has gone offline. For the solution, see [Case two: some DM-worker restarts](#case-two-some-dm-worker-restarts-or-is-unreachable-temporarily).

Author: @csuzhangxc Please share your thoughts. Thanks.

### Manual solution

1. Run `show-ddl-locks` to obtain the information of the sharding DDL lock that is currently pending synchronization. 

2. Run the `unlock-ddl-lock` command to specify the information of the lock to be unlocked manually.

    - If the owner of this lock is offline, you can configure the `--owner` parameter to specify another DM-worker as the new owner to execute the DDL operation.

3. Run `show-ddl-locks` to check whether this lock has been successfully unlocked.

### Impact

After you have manually unlocked the lock, the lock might not be automatically synchronized when the next sharding DDL event is received, because the configuration information of this task still includes offline DM-workers.

> Therefore, after you have manually unlocked the DM-workers, you need to use `stop-task`/`start-task` and a new task configuration that does not include offline DM-workers to restart the task.

Reviewer 2: ```suggestion
Therefore, after you have manually unlocked the DM-workers, you need to use `stop-task`/`start-task` and the updated task configuration that does not include offline DM-workers to restart the task.
```

> **Note:** If the DM-workers that went offline go online after you run `unlock-ddl-lock`, it means: These DM-workers will synchronize the unlocked DDL operation again. (Other DM-workers that were not offline have synchronized the DDL operation.)
The DDL operation of these DM-workers will try to match the subsequent synchronized DDL operations of other DM-workers. A match error of synchronizing sharding DDL operations of different DM-workers might occur.

## Case two: some DM-worker restarts (or is unreachable temporarily)

> The unlocking operation process of a DM calling multiple DM-workers to execute/skip the sharding DDL operation and update the checkpoint is not atomic. Therefore, a possible case is that after the owner finishes the DDL operation, an non-owner restarts before other DM-workers skip this DDL operation. At this time, the lock information on other DM-masters has been removed but the DM-worker that performs the restart operation has not skipped the DDL operation or updated the checkpoint.

Reviewer 2: ```suggestion
The unlocking operation process of a DM calling multiple DM-workers to execute/skip the sharding DDL operation and update the checkpoint is not atomic. Therefore, a possible case is that after the owner finishes the DDL operation, a non-owner restarts before other DM-workers skip this DDL operation. At this time, the lock information on the DM-master has been removed but the DM-worker that performs the restart operation has not skipped the DDL operation or updated the checkpoint.
```

only one DM-master

    - Specifies the lock ID of the DDL operation to be unlocked (this ID can be obtained by `show-ddl-locks`)

- `owner`: 
    
    - Flag parameter, string, `--owner`, optional
    - If set, this value should correspond to a DM-worker that executes the DDL operation of the lock instead of the default owner.

- `force-remove`: 
    
    - Flag parameter, boolean, `--force-remove`, optional
    - If set, the lock information is removed even though the owner fails to execute the DDL operation. The owner cannot retry or perform other operations.

- `worker`: 
    
    - Flag parameter, string array, `--worker`, optional
    - Can be specified repeatedly multiple times.
    - If not set, all the DM-workers to receive the lock event execute/skip the DDL operation. If set, only the specified DM-workers execute/skip the DDL operation.

#### `break-ddl-lock`

- `task-name`: 
    
    - Non-flag parameter, string, required
    - Specifies the name for the task that contains the lock to be broken.

- `worker`: 

    - Flag parameter, string, `--worker`, required
    - Only one should be specified.
    - Specifies the DM-worker to break the lock.

- `remove-id`: 

>     - Flag parameter, string, `--remove-id`, required

Reviewer 3: ```suggestion
    - Flag parameter, string, `--remove-id`, optional
```

    - If set, the value should be the ID of a DDL lock. Then the information about the DDL lock recorded in the DM-worker is removed.
    
- `exec`: 

    - Flag parameter, boolean, `--exec`, optional

>     - If set, a specific DM-worker executes the DDL operation corresponding to the lock.

Reviewer 3: ```suggestion
    - If set, a specific DM-worker executes the DDL statement corresponding to the lock.
```

>     - Cannot be specified simultaneously with the `skip` parameter.

Reviewer 3: ```suggestion
    - You cannot specify `exec` and `skip` at the same time.
```

- `skip`: 
    
    - Flag parameter, boolean, `--skip`, optional
    - If set, a specific DM-worker skips the DDL operation of the lock.

>     - Cannot be specified simultaneously with the `skip` parameter.

Reviewer 3: ```suggestion
    - You cannot specify `exec` and `skip` at the same time.
```
